### PR TITLE
Implement async socket introspection

### PR DESCRIPTION
**Pull request by Kaiepi**
This will make it possible to get the native descriptor of
IO::Socket::Async instances later on.

### CODE REVIEW

File: src/io/asyncsocket.c [hunk 1]
```diff
@@ -497,6 +497,105 @@ static void push_name_and_port(MVMThreadContext *tc, struct sockaddr_storage *na
...
+    MVMAsyncTask *task;
+    SocketInfo   *si;
+
+    /* Validate REPRS. */
+    if (REPR(queue)->ID != MVM_REPR_ID_ConcBlockingQueue)
+        MVM_exception_throw_adhoc(tc,
+            "asyncsocket target queue must have ConcBlockingQueue REPR");
+    if (REPR(async_type)->ID != MVM_REPR_ID_MVMAsyncTask)
+        MVM_exception_throw_adhoc(tc,
+            "asyncsocket result type must have REPR AsyncTask");
```
Comment (MasterDuke17) [inline src/io/asyncsocket.c]: Nitpick, but the first error message is '... type REPR' and the second is '... REPR type'.

File: src/io/asyncsocketudp.c [hunk 2]
```diff
@@ -443,10 +449,33 @@ static MVMint64 close_socket(MVMThreadContext *tc, MVMOSHandle *h) {
...
+    return (MVMint64)(handle->type == UV_TTY);
+}
+
+static MVMint64 socket_handle(MVMThreadContext *tc, MVMOSHandle *h) {
+    MVMIOAsyncUDPSocketData *data = (MVMIOAsyncUDPSocketData *)h->body.data;
+    uv_handle_t *handle = (uv_handle_t *)data->handle;
+    int fd;
+    uv_os_fd_t fh;
+    uv_fileno(handle, &fh);
+#ifdef _WIN32
```
Comment (ugexe) [inline src/io/asyncsocketudp.c]: See comment in `src/io/asyncsocket.c`

File: src/io/asyncsocket.c [hunk 3]
```diff
@@ -497,6 +526,102 @@ static void push_name_and_port(MVMThreadContext *tc, struct sockaddr_storage *na
...
+    if (REPR(queue)->ID != MVM_REPR_ID_ConcBlockingQueue)
+        MVM_exception_throw_adhoc(tc,
+            "asyncsocket target queue must have ConcBlockingQueue REPR (got %s)",
+             MVM_6model_get_stable_debug_name(tc, queue->st));
+    if (REPR(async_type)->ID != MVM_REPR_ID_MVMAsyncTask)
+        MVM_exception_throw_adhoc(tc,
+            "asyncsocket result type must have AsyncTask REPR");
+
+    /* Create async task handle. */
+    MVMROOT(tc, queue, {
```
Comment (ugexe) [inline src/io/asyncsocket.c]: Why `MVMROOT` only `queue` and not `schedulee`?

File: src/io/asyncsocket.c [hunk 4]
```diff
@@ -497,6 +526,102 @@ static void push_name_and_port(MVMThreadContext *tc, struct sockaddr_storage *na
...
+            MVM_repr_push_o(tc, arr, msg_box);
+        });
+        MVM_repr_push_o(tc, t->body.queue, arr);
+
+        /* Cleanup handles. */
+        uv_close((uv_handle_t *)si->socket, free_on_close_cb);
+        si->socket = NULL;
+        MVM_io_eventloop_remove_active_work(tc, &si->work_idx);
+    } else {
+        MVMROOT2(tc, arr, t, {
```
Comment (ugexe) [inline src/io/asyncsocket.c]: Why `MVMROOT` `t` here but not in the above conditional branch?

Comment (Kaiepi) [reply src/io/asyncsocket.c]: Missed that

File: src/io/asyncsocket.c (hunk 3 above)
Comment (Kaiepi) [reply src/io/asyncsocket.c]: Why would I need to include `schedulee` here?

Comment (Kaiepi) [reply src/io/asyncsocket.c]: Sorry, misread. Should `queue` be included here as well?

Comment (ugexe) [reply src/io/asyncsocket.c]: I think you want to root `queue` and `schedulee` -- there is an allocation on line 609, and then lines 611 and 612 gc collectable objects are passed.

**Comment by Kaiepi** [comment]
Updated

File: src/io/asyncsocket.c [hunk 5]
```diff
@@ -732,10 +851,8 @@ static void listen_setup(MVMThreadContext *tc, uv_loop_t *loop, MVMObject *async
     li->work_idx   = MVM_io_eventloop_add_active_work(tc, async_task);
 
     /* Create and initialize socket and connection, and start listening. */
```
Comment (ugexe) [inline src/io/asyncsocket.c]: Is this comment still correct that it is initializing a socket?

Comment (Kaiepi) [reply src/io/asyncsocket.c]: No

File: src/io/asyncsocket.c [hunk 6]
```diff
@@ -497,6 +515,102 @@ static void push_name_and_port(MVMThreadContext *tc, struct sockaddr_storage *na
...
+    si->work_idx      = MVM_io_eventloop_add_active_work(tc, async_task);
+
+    /* Create and initialize socket. */
+    si->socket       = MVM_malloc(sizeof(uv_tcp_t));
+    si->socket->data = data;
+    if ((r = uv_tcp_init(loop, si->socket)) < 0) {
+        MVMROOT2(tc, arr, t, {
+            MVM_repr_push_o(tc, arr, t->body.schedulee);
+            MVM_repr_push_o(tc, arr, tc->instance->boot_types.BOOTIO);
+            MVMString *msg_str = MVM_string_ascii_decode_nt(tc, tc->instance->VMString, uv_strerror(r));
```
Comment (ugexe) [inline src/io/asyncsocket.c]: ISO C90 forbids mixed declarations and code

Comment (Kaiepi) [reply src/io/asyncsocket.c]: Should be fixed now

**Pull request merged**